File: utils/queue.js
class Queue {
  constructor (arr = []) {
    this.arr = arr
  }

  empty () {
    return this.arr.length === 0
  }

  push (newElement) {
    this.arr.push(newElement)
  }

  pop () {
    return this.arr.shift()
  }

  front () {
    return this.arr[0]
  }

  back () {
    return this.arr[this.arr.length - 1]
  }
}

module.exports = Queue

File: problems/933-number-of-recent-calls/queue.cpp
/**
 * 队列：维护一个可用队列
 * 
 * 时间：172ms
 * 空间：`O(M) <= O(3000)`, 50.1MB
 */

class RecentCounter {
public:
    // O(1)
    RecentCounter() {}
    
    // 最坏情况是O(M)
    int ping(int t) {
        Q.push(t);

        // 删除不可能再有用的头部
        while (Q.front() < t - 3000) {
            Q.pop();
        }

        return Q.size();
    }

private:
    queue<int> Q;
};